{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = {'nt_muts': 'results/nt_muts.json',\n",
    "         'aa_muts': 'results/aa_muts.json',\n",
    "         'tree': 'results/tree.nwk',\n",
    "         'couplings': 'results/annotated_ranked_couplings.tsv',\n",
    "         'time': 'results/branch_lengths.json'}\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "sns.set_style('white')\n",
    "\n",
    "sns.set_context('talk')\n",
    "\n",
    "import json\n",
    "import itertools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import Phylo\n",
    "\n",
    "t = Phylo.read(files['tree'], 'newick')\n",
    "ngenomes = len({x.name for x in t.get_terminals() + t.get_nonterminals()\n",
    "                if 'NODE_' not in x.name})\n",
    "\n",
    "nt = json.load(open(files['nt_muts']))\n",
    "\n",
    "nt_muts = {k: {x: 1 for x in nt['nodes'][k]['muts']}\n",
    "           for k in nt['nodes']\n",
    "           if not k.startswith('NODE_')}\n",
    "ntn_muts = {k: {int(x[1:-1]): 1 for x in nt['nodes'][k]['muts']}\n",
    "           for k in nt['nodes']\n",
    "           if not k.startswith('NODE_')}\n",
    "\n",
    "for c in t.get_nonterminals():\n",
    "    if c.name not in nt['nodes']:\n",
    "        continue\n",
    "    for m in nt['nodes'][c.name]['muts']:\n",
    "        for n in c.get_terminals():\n",
    "            nt_muts[n.name][m] = 1\n",
    "            ntn_muts[n.name][int(m[1:-1])] = 1\n",
    "\n",
    "nt = {x for k in nt_muts for x in nt_muts[k]}\n",
    "ntn = {int(x[1:-1]) for x in nt}\n",
    "\n",
    "aa = json.load(open(files['aa_muts']))\n",
    "\n",
    "aa_muts = {}\n",
    "for sample in aa['nodes']:\n",
    "    if sample.startswith('NODE_'):\n",
    "        continue\n",
    "    aa_muts[sample] = {}\n",
    "    for gene, muts in aa['nodes'][sample]['aa_muts'].items():\n",
    "        if len(muts) > 0:\n",
    "            aa_muts[sample][gene] = set()\n",
    "        for mut in muts:\n",
    "            aa_muts[sample][gene].add(mut)\n",
    "\n",
    "for c in t.get_nonterminals():\n",
    "    if c.name not in aa['nodes']:\n",
    "        continue\n",
    "    for g, ms in aa['nodes'][c.name]['aa_muts'].items():\n",
    "        for n in c.get_terminals():\n",
    "            aa_muts[n.name][g] = aa_muts[n.name].get(g, set())\n",
    "            for m in ms: \n",
    "                aa_muts[n.name][g].add(m)\n",
    "\n",
    "aa = {'%s-%s' % (g, pos)\n",
    "      for k in aa_muts\n",
    "      for g in aa_muts[k]\n",
    "      for pos in aa_muts[k][g]}\n",
    "\n",
    "ntdf = pd.DataFrame(nt_muts)\n",
    "ntdf[np.isnan(ntdf)] = 0\n",
    "ntf = ntdf.T.sum() / ntdf.shape[1]\n",
    "ntndf = pd.DataFrame(ntn_muts)\n",
    "ntndf[np.isnan(ntndf)] = 0\n",
    "ntnf = ntndf.T.sum() / ntndf.shape[1]\n",
    "npos = set(ntf[ntf > 0.01].index)\n",
    "pos = {int(x[1:-1]) for x in set(ntf[ntf > 0.01].index)}\n",
    "npositions = len(npos)\n",
    "all_pos = {int(x[1:-1]) for x in set(ntf.index)}\n",
    "napositions = len(all_pos)\n",
    "\n",
    "m = pd.read_csv(files['couplings'],\n",
    "                sep='\\t')\n",
    "m = m[(m['pos1'].isin(pos)) &\n",
    "      (m['pos2'].isin(pos)) &\n",
    "      (m['direct'] == 1)]\n",
    "m = m.sort_values('coupling_strength')\n",
    "\n",
    "cpos = set(m[['pos1', 'pos2']].values.flatten()).intersection(ntnf[ntnf >= 0.01].index)\n",
    "nsites = len(cpos)\n",
    "ccpos = set(m[m['both_coding']][['pos1', 'pos2']].values.flatten()).intersection(ntnf[ntnf >= 0.01].index)\n",
    "ncsites = len(ccpos)\n",
    "\n",
    "bl = json.load(open(files['time']))\n",
    "tl = {k: bl['nodes'][k]['numdate']\n",
    "      for k in bl['nodes']\n",
    "      if not k.startswith('NODE_')}\n",
    "tl = pd.Series(tl)\n",
    "\n",
    "ntt = None\n",
    "trange = np.arange(tl.min(), tl.max(), 1 / 365)\n",
    "for t1 in trange:\n",
    "    if t1 + 20 / 365 > tl.max():\n",
    "        continue\n",
    "    samples = ntndf.columns.intersection(set(tl[(tl >= t1) & (tl < t1 + 20 / 365)].index))\n",
    "    tmp = ntndf.loc[:, samples].T.sum() / len(samples)\n",
    "    tmp.name = t1 + 20/365\n",
    "    if ntt is None:\n",
    "        ntt = tmp.to_frame()\n",
    "    else:\n",
    "        ntt = ntt.join(tmp.to_frame(), how='outer')\n",
    "ntt = ntt.stack().reset_index()\n",
    "ntt.columns = ['pos', 'time', 'af']\n",
    "\n",
    "tal = {k: bl['nodes'][k]['numdate']\n",
    "       for k in bl['nodes']}\n",
    "tacl = {k: bl['nodes'][k]['num_date_confidence']\n",
    "        for k in bl['nodes']}\n",
    "\n",
    "nt = json.load(open(files['nt_muts']))\n",
    "\n",
    "nt_time = {}\n",
    "nt_conf = {}\n",
    "for k in nt['nodes']:\n",
    "    time = tal[k]\n",
    "    time_c = tacl[k]\n",
    "    for x in nt['nodes'][k]['muts']:\n",
    "        curr_time = nt_time.get(int(x[1:-1]), np.nan)\n",
    "        if not np.isnan(curr_time) and curr_time < time:\n",
    "            continue\n",
    "        nt_time[int(x[1:-1])] = time\n",
    "        nt_conf[int(x[1:-1])] = time_c\n",
    "\n",
    "nt_time = pd.Series(nt_time)\n",
    "nt_time.name = 'time'\n",
    "\n",
    "nt_time = nt_time.to_frame().join(ntnf.to_frame(), how='outer')\n",
    "nt_time.columns = ['time', 'vf']\n",
    "nt_time.index.name = 'position'\n",
    "nt_time['Position'] = [x if x in cpos\n",
    "                       else 'Other'\n",
    "                       for x in nt_time.index]\n",
    "nt_time['tstart'] = [nt_conf[x][0]\n",
    "                     for x in nt_time.index]\n",
    "nt_time['tend'] = [nt_conf[x][1]\n",
    "                     for x in nt_time.index]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No signs of co-evolution between sites in the SARS-CoV-2 genomes sequenced to date, using Direct Coupling Analysis (DCA)\n",
    "=====\n",
    "\n",
    "[Marco Galardini and Ahmad S. Khalil](http://www.bu.edu/khalillab/)\n",
    "\n",
    "*Biological Design Center, Boston University, Boston, MA, USA*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Markdown as md\n",
    "\n",
    "import datetime\n",
    "today = datetime.datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "md(f\"**Last update:** {today}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5>To view the code used to generate the plots in this report <a href=\"javascript:toggle()\" target=\"_self\"><b>click here</b></a>. The whole pipeline can instead be accessed <a href=\"https://github.com/khalillab/ncov-dca\">here</a></h5>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Introduction\n",
    "---\n",
    "\n",
    "The current COVID-19 pandemic is accelerating the advent of genomic epidemiology,\n",
    "with tens of genomes of the SARS-CoV-2 virus made available every day.\n",
    "This allows a near real-time interpretation of viral genetic data, aimed at understanding\n",
    "the evolutionary origin of the virus, transmission chains, diagnostic strategies,\n",
    "and targets for pharmacological interventions.\n",
    "Among evolutionary approaches, the study of co-evolution can highlight\n",
    "sites in the genome that interact directly or indirectly. Recent methodological advances allow\n",
    "to apply Direct Coupling Analysis (DCA) across whole genomes and thousands of samples to highlight\n",
    "co-evolving sites. Here we apply DCA to {ngenomes} whole SARS-Cov-2 genome sequences, and although\n",
    "the method indicates that {nsites} are strongly coupled, we argue that this is likely due\n",
    "to these sites mutating early in the pandemic and thus having a high frequency in the population.\n",
    "Moreover, we observe no sign of convergence\n",
    "(*i.e.* strongly coupled sites co-mutating in multiple clades) which further reinforces\n",
    "our conclusion that thus far no co-evolving sites can be used to infer direct and indirect\n",
    "interactions betweeen variant sites.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Results\n",
    "---\n",
    "\n",
    "Using the ncov [nextstrain](https://academic.oup.com/bioinformatics/article/34/23/4121/5001388) [pipeline](https://github.com/nextstrain/ncov)\n",
    "we have generated a whole genome alignment of {ngenomes} samples, obtained\n",
    "from the GISAID database on {today} (see the **acknowledgments** section for\n",
    "information about the labs who kindly submitted sequences). Based on this\n",
    "alignment we observe {npositions} sites with variant frequency above 1% (**Figure 1**).''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ntmat = ntndf.loc[ntndf.index.intersection(pos)].sort_index()\n",
    "\n",
    "# Sort the matrix according to tip labels in the tree\n",
    "nt_sorted = ntmat[[x.name for x in t.get_terminals()]]\n",
    "\n",
    "# Plot presence/absence matrix against the tree\n",
    "with sns.axes_style('whitegrid'):\n",
    "    fig = plt.figure(figsize=(18, 10))\n",
    "\n",
    "    ax1=plt.subplot2grid((1,40), (0, 5), colspan=35)\n",
    "    a=ax1.matshow(nt_sorted.T, cmap=plt.cm.Blues_r,\n",
    "               vmin=0, vmax=1,\n",
    "               aspect='auto',\n",
    "               interpolation='none',\n",
    "                )\n",
    "    ax1.set_yticks([])\n",
    "    ax1.set_xticks([])\n",
    "    ax1.set_xticks([i for i, x in enumerate(nt_sorted.index)])\n",
    "    ax1.set_xticklabels([str(x) for x in nt_sorted.index],\n",
    "                        rotation=90,\n",
    "                        fontdict={'color': 'k'})\n",
    "    for tick in ax1.get_xticklabels():\n",
    "        if int(tick.get_text()) in cpos:\n",
    "            tick.set_color('r')\n",
    "    ax1.tick_params(grid_linewidth=0, labelsize=12, direction='out', length=0)\n",
    "    ax1.set_xlabel('Positions')\n",
    "    ax1.set_ylabel('Samples')\n",
    "    ax1.yaxis.set_label_position('right')\n",
    "    for i, x in enumerate(nt_sorted.index):\n",
    "        ax1.axvline(i - 0.5, color='white', linewidth=0.66)\n",
    "    sns.despine(ax=ax1, left=True, bottom=True)\n",
    "\n",
    "    ax = fig.add_subplot(1,2,1)\n",
    "    # matplotlib v1/2 workaround\n",
    "    try:\n",
    "        ax=plt.subplot2grid((1,40), (0, 0), colspan=5, facecolor='white')\n",
    "    except AttributeError:\n",
    "        ax=plt.subplot2grid((1,40), (0, 0), colspan=5, axisbg='white')\n",
    "\n",
    "    fig.subplots_adjust(wspace=0, hspace=0)\n",
    "\n",
    "    with plt.rc_context({'lines.linewidth': 0.5}):\n",
    "        Phylo.draw(t, axes=ax, \n",
    "                   show_confidence=False,\n",
    "                   label_func=lambda x: None,\n",
    "                   xticks=([],), yticks=([],),\n",
    "                   ylabel=('',), xlabel=('',),\n",
    "                   xlim=(0.0009,0.0013),\n",
    "                   axis=('off',),\n",
    "                   do_show=False,\n",
    "                  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''_**Figure 1.** SARS-CoV-2 time-corrected phylogenetic tree and variant sites presence/absence matrix.\n",
    "In the matrix each row represent a viral genome and each column a position in the genome with at least 1%\n",
    "of variant sites across all samples. White represents the presence of a variant with respect to the\n",
    "reference genome (GenBank: [MN908947](https://www.ncbi.nlm.nih.gov/nuccore/MN908947)), blue absence of variants.\n",
    "Columns are ordered based on their position, from left to right. The {nsites} coupled positions are highlighted in red._''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Next, we applied Direct Coupling Analysis to sites with variant frequency above 1%,\n",
    "using the implementation provided by [SpydrPick](https://academic.oup.com/nar/article/47/18/e112/5541093).\n",
    "We found {nsites} to be strongly coupled over {npositions} used as input, meaning that according\n",
    "to this method roughly half of the positions with higher frequency variants\n",
    "are co-evolving (**Figure 2**). **We however argue that this result is\n",
    "due to two confounding factors, which suggest that no molecular or functional conclusions can be made at this stage.**\n",
    "First, as can be clearly observed in **Figure 1**, virtually all variants are exclusive to one clade,\n",
    "with no signs of convergence. If coupled sites were found to be variant across multiple clades, that\n",
    "would be a stronger indication of co-evolution, since the likelihood of observing two sites co-mutating\n",
    "in independent lineages in absence of recombination is very small. Second, all coupled variants appeared\n",
    "early in the spread of the pandemic, as estimated by the ancestral sequence reconstruction by \n",
    "maximum likelihood implemented in [TreeTime](https://academic.oup.com/ve/article/4/1/vex042/4794731) (**Figure 3**).\n",
    "This observation implies that coupled mutations might have risen to high frequency only due to chance, and not\n",
    "because they confer a selective advantage alone or due to their direct/indirect interaction.\n",
    "Furthermore, we observe very little change in the observed frequencies of these\n",
    "variants over time (**Figure 4**) which might also be an indication that none of them\n",
    "is more likely to be adaptive.\n",
    "Given these results, we are not showing further analysis about the position of\n",
    "these {nsites} high frequency variants in relation to the genes encoded in the virus'\n",
    "genome, so that there is a reduced risk of misinterpreting these results.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(7, 3))\n",
    "sns.boxplot(data=[ntnf.loc[cpos], ntnf.loc[ntnf.index.difference(cpos)]],\n",
    "            orient='h', palette=[sns.xkcd_rgb['orange red'],\n",
    "                                 sns.xkcd_rgb['light grey']])\n",
    "plt.axvline(0.01, color='r', ls='dashed', label='VF threshold')\n",
    "plt.yticks([0, 1], ['Coupled', 'Uncoupled or low frequency'])\n",
    "plt.xlabel('Variant frequency')\n",
    "plt.legend(loc='lower right',\n",
    "           bbox_to_anchor=(1, 0));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''_**Figure 2.** Variant frequency of all observed variants (N={napositions}).\n",
    "About half of frequent variants are found to be coupled by DCA ({nsites}/{npositions})_''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "months = {2020 - 15/365: '2019-12-15',\n",
    "          2020.0: '2020-01-01',\n",
    "          2020 + 15/366: '2020-01-15',\n",
    "          2020 + 32/366: '2020-02-01',\n",
    "          2020 + 32/366 + 15/366: '2020-02-15',\n",
    "          2020 + 32/366 + 30/366: '2020-03-01',\n",
    "          2020 + 32/366 + 30/366 + 15/366: '2020-03-15',\n",
    "          2020 + 32/366 + 30/366 + 32/366: '2020-04-01'}\n",
    "\n",
    "rel = sns.relplot(x='time', y='vf',\n",
    "            hue='Position', data=nt_time,\n",
    "            hue_order=['Other'] + list(nt_time[nt_time['Position'] != 'Other'].index),\n",
    "            palette=[sns.xkcd_rgb['light grey']] + list(sns.color_palette('hsv',\n",
    "                                           len(nt_time[nt_time['Position'] != 'Other'].index))),\n",
    "            height=5, aspect=2)\n",
    "for x, vf in nt_time[nt_time['Position'] != 'Other'][['vf']].iterrows():\n",
    "    plt.plot([nt_conf[x][0], nt_conf[x][1]],\n",
    "             [vf, vf], '-', alpha=0.1, color='k', zorder=-1)\n",
    "plt.xlabel('Time of MRCA')\n",
    "plt.ylabel('Variant frequency\\n(across all samples)')\n",
    "plt.xticks(sorted(months),\n",
    "           [months[x] for x in sorted(months)],\n",
    "           rotation=45, ha='right', size=12)\n",
    "rel._legend.remove()\n",
    "plt.legend(fontsize=10,\n",
    "           ncol=2, title='',\n",
    "           loc='center left',\n",
    "           bbox_to_anchor=(1, 0.5))\n",
    "plt.ylim(0.999E-4, 0.6)\n",
    "plt.yscale('log');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''_**Figure 3.** Early variants are more frequently observed.\n",
    "Estimated age of each variant (N={napositions}) estimated through\n",
    "[TreeTime](https://academic.oup.com/ve/article/4/1/vex042/4794731) is compared against\n",
    "its observed frequency across all samples sequenced to date. The {nsites}\n",
    "coupled sites are highlighted. MRCA, Most Recent Common Ancestor.\n",
    "Horizontal lines indicate the confidence interval for the age\n",
    "of each coupled variant._''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "months = {2020 + 15/366: '2020-01-15',\n",
    "          2020 + 32/366: '2020-02-01',\n",
    "          2020 + 32/366 + 15/366: '2020-02-15',\n",
    "          2020 + 32/366 + 30/366: '2020-03-01',\n",
    "          2020 + 32/366 + 30/366 + 15/366: '2020-03-15',\n",
    "          2020 + 32/366 + 30/366 + 32/366: '2020-04-01'}\n",
    "\n",
    "lcolor = [sns.xkcd_rgb['light grey']] * len(ntt[np.invert(ntt['pos'].isin(cpos))]['pos'].unique())\n",
    "lcolor += sns.color_palette('hsv', len(ntt[ntt['pos'].isin(cpos)]['pos'].unique()))\n",
    "rel = sns.relplot(x='time',\n",
    "            y='af',\n",
    "            hue='pos',\n",
    "            kind='line',\n",
    "            data=ntt[ntt['af']!=0],\n",
    "            palette=lcolor,\n",
    "            hue_order=sorted(set(ntt['pos'].unique()).difference(cpos)) + sorted(cpos),\n",
    "            height=5, aspect=2)\n",
    "plt.xlabel('Time')\n",
    "plt.ylabel('Variant frequency\\n(at specific timepoint)')\n",
    "plt.xticks(sorted(months),\n",
    "           [months[x] for x in sorted(months)],\n",
    "           rotation=45, ha='right', size=12)\n",
    "handles, labels = rel.ax.get_legend_handles_labels()\n",
    "by_label = {x: y for x, y in zip(labels, handles)\n",
    "            if x != 'pos' and int(x) in cpos}\n",
    "rel._legend.remove()\n",
    "plt.legend(by_label.values(), by_label.keys(),\n",
    "           fontsize=10,\n",
    "           ncol=2, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "plt.ylim(0.999E-4, 0.65)\n",
    "plt.yscale('log');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''_**Figure 4.** Empirical change in variant frequencies over time.\n",
    "For each time point, only the samples from the previous 20 days are considered\n",
    "to compute variant frequencies. The {nsites} coupled sites are highlighted._''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Discussion\n",
    "---\n",
    "\n",
    "The current push to scale up genomic sequencing is already delivering tens of viral genomes\n",
    "released every day by a large network of laboratories across the world. Here, we focused\n",
    "on understanding whether the genomic data collected to date could be used to find evidence of\n",
    "co-evolution between sites along the SARS-CoV-2 genome. We have used a method based on Direct Coupling Analysis\n",
    "([SpydrPick](https://academic.oup.com/nar/article/47/18/e112/5541093))\n",
    "to highlight {nsites} sites observed in {ngenomes} samples derived from [GISAID](https://platform.gisaid.org/epi3/frontend)\n",
    "as of {today}, and pre-processed using the [nextstrain pipeline](https://academic.oup.com/bioinformatics/article/34/23/4121/5001388).\n",
    "Based on this analysis, **we believe that the sites found to be strongly coupled by DCA\n",
    "are in fact most likely false positives**, given the absence of convergence and the age\n",
    "of mutations at those sites. We however acknowledge that methods that explicitly consider\n",
    "the information about the age of each variant across the phylogeny might be able to overcome\n",
    "these problems, as shown in [a study for influenza A surface proteins](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1001301).\n",
    "Moreover, as the pandemics progresses in the months to come, we will revise this analysis with the hundreds of\n",
    "new genome sequences that will likely be sequenced.''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Methods\n",
    "---\n",
    "\n",
    "#### **Data preprocessing**\n",
    "\n",
    "We downloaded {ngenomes} SARS-CoV-2 genomes from GISAID on {today}, and used the\n",
    "[nextstrain ncov pipeline](https://github.com/nextstrain/ncov) to filter and align them to the the reference\n",
    "sequence from GenBank [MN908947](https://www.ncbi.nlm.nih.gov/nuccore/MN908947), using [mafft](https://academic.oup.com/mbe/article/30/4/772/1073398).\n",
    "[IQ-tree](https://academic.oup.com/mbe/article/32/1/268/2925592) was\n",
    "used to reconstruct the phylogeny, further refined by [TreeTime](https://academic.oup.com/ve/article/4/1/vex042/4794731).\n",
    "Maximum likelihood ancestral reconstruction as implemented\n",
    "in [TreeTime](https://academic.oup.com/ve/article/4/1/vex042/4794731) was used to\n",
    "infer the age of each observed variant.\n",
    "\n",
    "#### **Direct Coupling Analysis**\n",
    "\n",
    "We used the masked nucleotide sequence alignment as input for\n",
    "[SpydrPick](https://academic.oup.com/nar/article/47/18/e112/5541093)\n",
    "using an allele frequency filter at 1% and sequencing reweighting threshold of 1.\n",
    "Only pairs of sites estimated to\n",
    "be strongly coupled (mutual information > 0.11) are retained for further analysis.\n",
    "\n",
    "#### **Visualization**\n",
    "\n",
    "The data was post-processed using the [Python programming language](https://www.python.org/) and\n",
    "the jupyterlab environment, using the [numpy](https://aip.scitation.org/doi/abs/10.1109/MCSE.2011.37),\n",
    "[scipy](https://www.nature.com/articles/s41592-019-0686-2), [biopython](https://academic.oup.com/bioinformatics/article/25/11/1422/330687),\n",
    "[matplotlib](https://ieeexplore.ieee.org/document/4160265) and [seaborn](https://zenodo.org/record/3629446) libraries.\n",
    "\n",
    "#### **Reproducibility**\n",
    "\n",
    "The code used is available as [a git repository on github](https://github.com/khalillab/ncov-dca), which\n",
    "is an extension of the existing [nextstrain ncov pipeline](https://github.com/nextstrain/ncov).''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "md(f'''Acknowledgements\n",
    "---\n",
    "\n",
    "We are grateful to Dr. John Lees for providing feedback on implementations of DCA,\n",
    "and to the [nextstrain team](https://nextstrain.org/) for providing the community with a robust and open analysis pipeline.\n",
    "We are also grateful to all laboratories that submitted SARS-CoV-2 genomes to [GISAID](https://platform.gisaid.org/epi3/fronten),\n",
    "specifically:\n",
    "\n",
    "- AIDS Vaccine Research Laboratories\n",
    "- ARGO Open Lab Platform for Genome sequencing\n",
    "- African Centre of Excellence for Genomics of Infectious Diseases (ACEGID), Redeemer's University, Ede, Osun State, Nigeria\n",
    "- Alaska State Public Health Virology Laboratory\n",
    "- Andersen Lab, The Scripps Research Institute\n",
    "- BCCDC Public Health Laboratory\n",
    "- BGI & Institute of Microbiology, Chinese Academy of Sciences & Shandong First Medical University\n",
    "- Beijing Genomics Institute (BGI)\n",
    "- Beijing Institute of Microbiology and Epidemiology\n",
    "- Bergthaler laboratory, CeMM Research Center for Molecular Medicine of the Austrian Academy of Sciences\n",
    "- Bioinformatics Laboratory - LNCC\n",
    "- Bioinformatics Research Group, Szentagothai Research Centre, University of Pecs\n",
    "- Bundeswehr Institute of Microbiology\n",
    "- CNR Virus des Infections Respiratoires - France SUD\n",
    "- Center of Medical Microbiology, Virology, and Hospital Hygiene, University of Duesseldorf\n",
    "- Centers for Disease Control, R.O.C. (Taiwan)\n",
    "- Chan-Zuckerberg Biohub\n",
    "- Charite Universitaetsmedizin Berlin, Institute of Virology\n",
    "- Chiu Laboratory, University of California, San Francisco\n",
    "- Chongqing Municipal Center for Disease Control and Prevention\n",
    "- University of Melbourne at The Peter Doherty Institute for Infection and Immunity, and the Victorian Infectious Disease Reference Laboratory\n",
    "- Dasman Diabetes Institute\n",
    "- Department for Virology, Molecular Biology and Genome Research, R. G. Lugar Center for Public Health Research, National Center for Disease Control and Public Health (NCDC) of Georgia.\n",
    "- Department of Clinical Diagnostics\n",
    "- Department of Health Technology and Informatics, Faculty of Health and Social Science, The Hong Kong Polytechnic University\n",
    "- Department of Healthcare Biotechnology\n",
    "- Department of Infection, Immunity and Cardiovascular Disease, The Florey Institute, The Medical School, University of Sheffield\n",
    "- Department of Laboratory Medicine, Lin-Kou Chang Gung Memorial Hospital, Taoyuan, Taiwan\n",
    "- Department of Medical Microbiology, Faculty of Medicine, University of Malaya\n",
    "- Department of Microbiology, Faculty of Medicine, Chinese University of Hong Kong, Hong Kong SAR, China\n",
    "- Department of Microbiology, Guangdong Provincial Center for Diseases Control and Prevention\n",
    "- Department of Microbiology, Institute for Viral Diseases, College of Medicine, Korea University\n",
    "- Department of Microbiology, PathWest QEII Medical Centre\n",
    "- Department of Microbiology, Zhejiang Provincial Center for Disease Control and Prevention\n",
    "- Department of Pathology and Medicine, New York University School of Medicine\n",
    "- Department of Virology, Faculty of Medicine, University of Helsinki, Helsinki, Finland\n",
    "- Department of microbiology laboratory, Anhui Provincial Center for Disease Control and Prevention\n",
    "- Departments of Pathology and Medicine, New York University School of Medicine\n",
    "- Division of Viral Diseases, Center for Laboratory Control of Infectious Diseases, Korea Centers for Diseases Control and Prevention\n",
    "- Division of Viral Diseases, Centers for Disease Control and Prevention\n",
    "- Erasmus Medical Center\n",
    "- Fujian Center for Disease Control and Prevention\n",
    "- GIGA Medical Genomics\n",
    "- Geelong Centre for Emerging Infectious Diseases\n",
    "- Gorgas Memorial Institute for Health Studies\n",
    "- Grubaugh Lab - Yale School of Public Health\n",
    "- Guangdong Provincial Center for Disease Control and Prevention\n",
    "- Guangdong Provincial Institution of Public Health\n",
    "- Hangzhou Center for Disease and Control Microbiology Lab\n",
    "- Hong Kong Department of Health\n",
    "- Hopitaux universitaires de Geneve Laboratoire de Virologie\n",
    "- Hospital Universitari Vall d'Hebron (HUVH) - Vall d'Hebron Research Institute (VHIR)\n",
    "- Hospital Universitario La Paz\n",
    "- Hubei Provincial Center for Disease Control and Prevention\n",
    "- INMI Lazzaro Spallanzani IRCCS\n",
    "- Indian Council of Medical Research-National Institute of Virology, Microbial Containment Complex\n",
    "- Infectious Disease Control Center\n",
    "- Infectious Disease Research Department, King Abdullah International Medical Research Center (KAIMRC)\n",
    "- Inspection Center of Hanghzou Center for Disease Control and Prevention\n",
    "- Institut Pasteur de Dakar\n",
    "- Institute information KU Leuven, Clinical and Epidemiological Virology\n",
    "- Institute of Environmental Science and Research (ESR)\n",
    "- Institute of Medical Virology, University of Zurich\n",
    "- Institute of Microbiology and Immunology, Faculty of Medicine, University of Ljubljana\n",
    "- Institute of Microbiology, Universidad San Francisco de Quito\n",
    "- Institute of Pathogen Biology, Chinese Academy of Medical Sciences & Peking Union Medical College\n",
    "- Institute of Viral Disease Control and Prevention, China CDC\n",
    "- Institute of Virology, Biomedical Research Center of the Slovak Academy of Sciences, Bratislava; Comenius University Science Park, Bratislava\n",
    "- Instituto Adolfo Lutz, Interdisciplinary Procedures Center, Strategic Laboratory\n",
    "- Instituto Gulbenkian de Ciencia\n",
    "- Instituto Nacional Enfermedades Infecciosas C.G.Malbran\n",
    "- Instituto Nacional de Salud, Universidad Cooperativa de Colombia, Instituto Alexander von Humboldt, Imperial College-London, London School of Hygiene & Tropical Medicine\n",
    "- Instituto Nacional de Saude (INSA)\n",
    "- Instituto Oswaldo Cruz FIOCRUZ - Laboratory of Respiratory Viruses and Measles (LVRS)\n",
    "- Instituto de Diagnostico y Referencia Epidemiologicos (INDRE)\n",
    "- Instituto de Salud Carlos III\n",
    "- Instituto de Salud Publica de Chile\n",
    "- IrsiCaixa AIDS Research Lab\n",
    "- Israel Institute for Biological Research\n",
    "- Istituto Zooprofilattico Sperimentale dell'Abruzzo e del Molise \"G.Caporale\"\n",
    "- Jiangsu Provincial Center for Disease Control & Prevention\n",
    "- Jiangxi province Center for Disease Control and Prevention\n",
    "- KU Leuven, Clinical and Epidemiological Virology\n",
    "- Kabara Cancer Research Institute\n",
    "- Key Laboratory of Human Diseases, Comparative Medicine, Institute of Laboratory Animal Science\n",
    "- Key Laboratory of Medical Molecular Virology (MOE-NHC-CAMS)\n",
    "- Korea Centers for Disease Control & Prevention (KCDC) Center for Laboratory Control of Infectious Diseases Division of Viral Diseases\n",
    "- Laboratoire National de Sante, Microbiology, Epidemiology and Microbial Genomics\n",
    "- Laboratoire Nationale de Sante, Microbiology, Epidemiology and Microbial Genomics\n",
    "- Laboratoire Virpath, CIRI U111, UCBL1, INSERM, CNRS, ENS Lyon\n",
    "- Laboratorio de Ecologia de Doencas Transmissiveis na Amazonia, Instituto Leonidas e Maria Deane - Fiocruz Amazonia\n",
    "- Laboratorio de Referencia Nacional de Biotecnologia y Biologia Molecular. Instituto Nacional de Salud Peru\n",
    "- Laboratorio di Microbiologia e Virologia, Universita Vita-Salute San Raffaele, Milano\n",
    "- Laboratory for Functional Genome Analysis, Dept. Genomics, Gene Center of the LMU Munich\n",
    "- Laboratory of Biology, Department of Medicine, Democritus University of Thrace, Greece\n",
    "- Laboratory of Infectious Diseases, Department of Biomedical and Clinical Sciences L. Sacco, University of Milan\n",
    "- Laboratory of Virology, INMI Lazzaro Spallanzani IRCCS\n",
    "- Li Ka Shing Faculty of Medicine, The University of Hong Kong\n",
    "- MRC-University of Glasgow Centre for Virus Research\n",
    "- MSHS Pathogen Surveillance Program\n",
    "- Malaysia Genome Institute\n",
    "- Microbial Genomics Core Lab, National Taiwan University Centers of Genomic and Precision Medicine\n",
    "- Microbiological Diagnostic Unit Public Health Laboratory\n",
    "- Microbiology and Immunology department\n",
    "- Ministry of Health Turkey\n",
    "- Minnesota Department of Health, Public Health Laboratory\n",
    "- Molecular Biology and Biotechnology Lab II\n",
    "- NSW Health Pathology - Institute of Clinical Pathology and Medical Research; Westmead Hospital; University of Sydney\n",
    "- National Centre for Infectious Diseases, National Centre for Infectious Diseases\n",
    "- National Influenza Center - National Institute of Hygiene and Epidemiology (NIHE)\n",
    "- National Influenza Center, Indian Council of Medical Research - National Institute of Virology\n",
    "- National Influenza Center, National Institute of Hygiene and Epidemiology (NIHE)\n",
    "- National Institute for Communicable Disease Control and Prevention (ICDC) Chinese Center for Disease Control and Prevention (China CDC)\n",
    "- National Institute for Communicable Diseases of the National Health Laboratory Service\n",
    "- National Institute for Viral Disease Control and Prevention, China CDC\n",
    "- National Microbiology Laboratory\n",
    "- National Public Health Laboratory\n",
    "- National Public Health Laboratory, National Centre for Infectious Diseases\n",
    "- National Reference Center for Viruses of Respiratory Infections, Institut Pasteur, Paris\n",
    "- National Research Center for Translational Medicine (Shanghai), Ruijin Hospital affiliated to Shanghai Jiao Tong University School of Medicine & Shanghai Public Health Clinical Center\n",
    "- Norwegian Institute of Public Health, Department of Virology\n",
    "- Onderzoeksgroep Virologie\n",
    "- Ontario Agency for Health Protection and Promotion (OAHPP)\n",
    "- PathWest Laboratory Medicine WA\n",
    "- Pathogen Discovery, Respiratory Viruses Branch, Division of Viral Diseases, Centers for Disease Control and Prevention\n",
    "- Pathogen Genomics Center, National Institute of Infectious Diseases\n",
    "- Pathogen Sequencing Lab, National Institute for Biomedical Research (INRB)\n",
    "- Programme in Emerging Infectious Diseases, Duke-NUS Medical School\n",
    "- Public Health Laboratory, Saudi CDC\n",
    "- Public Health Ontario Laboratories\n",
    "- Public Health Virology Laboratory\n",
    "- Public Health Wales Microbiology Cardiff\n",
    "- R. G. Lugar Center for Public Health Research, National Center for Disease Control and Public Health (NCDC) of Georgia.\n",
    "- Respiratory Virus Unit, Microbiology Services Colindale, Public Health England\n",
    "- Ryota Kumagai Tokyo Metropolitan Institute of Public Health\n",
    "- SCSFRI, South China Sea Fisheries Research Institute, Chinese Academy of Fishery Sciences (SCSFRI, CAFS)\n",
    "- Saitama Medical University\n",
    "- School of Public Health, The University of Hon g Kong\n",
    "- School of Public Health, The University of Hong Kong\n",
    "- Seattle Flu Study\n",
    "- Second Hospital of Anhui Medical University\n",
    "- Sequencing and Bioinformatics Service and Molecular Epidemiology Research Group. FISABIO-Public Health\n",
    "- Shenzhen Key Laboratory of Pathogen and Immunity, National Clinical Research Center for Infectious Disease, Shenzhen Third People's Hospital\n",
    "- South China Agricultural University\n",
    "- State Key Laboratory for Diagnosis and Treatment of Infectious Diseases, National Clinical Research Center for Infectious Diseases, First Affiliated Hospital, Zhejiang University School of Medicine, Hangzhou, China\n",
    "- State Key Laboratory of Virology, Wuhan University\n",
    "- State Veterinary Institute Prague\n",
    "- Statens Serum Institute\n",
    "- Swiss National Reference Centre for Influenza\n",
    "- Taiwan Centers for Disease Control\n",
    "- Takayuki Hishiki Kanagawa Prefectural Institute of Public Health, Department of Microbiology\n",
    "- Technology Centre, Guangzhou Customs\n",
    "- The First Affiliated Hospital of Guangzhou Medical University & BGI-Shenzhen\n",
    "- The University of Hong Kong\n",
    "- UCD National Virus Reference Laboratory\n",
    "- UW Virology Lab\n",
    "- Unit for Laboratory Development and Technology Transfer, Public Health Agency of Sweden\n",
    "- University Hospital Basel, Clinical Bacteriology\n",
    "- University Hospital Basel, Clinical Virology\n",
    "- University Hospital Basel, Labormedizin\n",
    "- University of California, San Francisco\n",
    "- University of Otago\n",
    "- University of Washington Virology Lab\n",
    "- University of Wisconsin - Madison AIDS Vaccine Research Laboratories\n",
    "- Utah Public Health Laboratory\n",
    "- VA DCLS\n",
    "- ViFU\n",
    "- Victorian Infectious Diseases Reference Laboratory and Microbiological Diagnostic Unit Public Health Laboratory, Doherty Institute\n",
    "- Virginia Division of Consolidated Laboratories\n",
    "- ViroGenetics - BSL3 Laboratory of Virology; Human Genome Variation Research Group & Genomics Centre MCB; Bioinformatics Research Group Department of Virology\n",
    "- Virology Department, Royal Infirmary of Edinburgh, NHS Lothian\n",
    "- Virology Laboratory, Scientific Department, Army Medical Center\n",
    "- Virology Unit, Institut Pasteur du Cambodge\n",
    "- Virology and Legal Medicine Laboratories, Department of Biomedical Sciences and Public Health, University Politecnica delle Marche\n",
    "- WHO National Influenza Centre Russian Federation\n",
    "- Weifang Center for Disease Control and Prevention & BGI-Shenzhen\n",
    "- Wuhan Institute of Virology, Chinese Academy of Sciences\n",
    "- deCODE genetics\n",
    "- the First Affiliated Hospital of Guangzhou Medical University & BGI-Shenzhen\n",
    "- Department of Medical Sciences, Ministry of Public Health, Thailand\n",
    "- Thai Red Cross Emerging Infectious Diseases - Health Science Centre\n",
    "- Department of Disease Control, Ministry of Public Health, Thailand\n",
    "- Shandong Academy of Medical Sciences\n",
    "- General Hospital of Central Theater Command of People's Liberation Army of China\n",
    "- Imperial College-London London School of Hygiene & Tropical Medicine''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ack = pd.read_csv('data/metadata.tsv', sep='\\t')\n",
    "\n",
    "# for i in sorted({x.replace('  ', ' ').replace('/', '-') for x in ack[['submitting_lab']].values.flatten()}):\n",
    "#     print('- ' + i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
